File: usb_camera/src/video_capture.cpp
#include <ros/ros.h>
#include <sensor_msgs/image_encodings.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>


using namespace cv;
using namespace std;


//--------------------------- Global variable ---------------------------//
ros::Publisher img_pub;
ros::Publisher caminfo_pub;
sensor_msgs::ImagePtr img_msg;
sensor_msgs::CameraInfo caminfo_msg;

// parameter
string save_folder;
bool show_flag;
bool save_flag;
string serial_port;
bool publish_camera_info;


//--------------------------- Internal funciton ---------------------------//
void GetCameraInfo(ros::NodeHandle nh_priv, sensor_msgs::CameraInfo& cam_info)
{
    string frame_id, distortion_model;
    int height, width, binning_x, binning_y;
    int roi_x_offset, roi_y_offset, roi_height, roi_width;
    vector<double> D, K, R, P;
    bool roi_do_rectify;

    nh_priv.param<string>("frame_id", frame_id, "camera"); 
    nh_priv.param<int>("height", height, 0); 
    nh_priv.param<int>("width", width, 0);    
    nh_priv.param<string>("distortion_model", distortion_model, "");    
    nh_priv.getParam("D", D);
    nh_priv.getParam("K", K);
    nh_priv.getParam("R", R);
    nh_priv.getParam("P", P);
    nh_priv.param<int>("binning_x", binning_x, 0);   
    nh_priv.param<int>("binning_y", binning_y, 0);       
    nh_priv.param<int>("roi/x_offset", roi_x_offset, 0);  
    nh_priv.param<int>("roi/y_offset", roi_y_offset, 0);  
    nh_priv.param<int>("roi/height", roi_height, 0);  
    nh_priv.param<int>("roi/width", roi_width, 0);  
    nh_priv.param<bool>("roi/do_rectify", roi_do_rectify, false);  

    cam_info.header.frame_id = frame_id;
    cam_info.height = height;
    cam_info.width = width;
    cam_info.distortion_model = distortion_model;
    cam_info.D = D;
    memcpy(&cam_info.K[0],  &K[0],  cam_info.K.size() * sizeof(double));
    memcpy(&cam_info.R[0],  &R[0],  cam_info.R.size() * sizeof(double));
    memcpy(&cam_info.P[0],  &P[0],  cam_info.P.size() * sizeof(double));
  
    cam_info.binning_x = binning_x;
    cam_info.binning_y = binning_y;
    cam_info.roi.x_offset = roi_x_offset;
    cam_info.roi.y_offset = roi_y_offset;
    cam_info.roi.height = roi_height;
    cam_info.roi.width = roi_width;
    cam_info.roi.do_rectify = roi_do_rectify;
}


//--------------------------- Main funciton ---------------------------//
int main(int argc, char * argv[])
{
    // Initial node
    ros::init(argc, argv, "video_capture");
    ros::NodeHandle nh;
    ros::NodeHandle nh_priv("~");
    ROS_INFO("Start the video_capture node\n");

    // Parameter
    nh_priv.param<string>("save_folder", save_folder, ""); 
    nh_priv.param<bool>("show_flag", show_flag, false); 
    nh_priv.param<bool>("save_flag", save_flag, false);  
    nh_priv.param<string>("serial_port", serial_port, "/dev/ttyVideo0"); 
    nh_priv.param<bool>("publish_camera_info", publish_camera_info, false);  

    // Camera parameter
    if(publish_camera_info)
    {
        GetCameraInfo(nh_priv, caminfo_msg);
        caminfo_pub = nh.advertise<sensor_msgs::CameraInfo>("camera_info", 1);
    }

    // Prepare to publish image topic
    image_transport::ImageTransport it(nh);
    image_transport::Publisher pub = it.advertise("image", 1);
    sensor_msgs::ImagePtr msg;

    // Open VideoCapture
    VideoCapture cap(serial_port);
    if(!cap.isOpened())
    {
        ROS_ERROR("Read video frame failed!\n");
        return 0;
    }

    // Read image
    Mat img;
    int count = 0;
    char imgname[128];
    while(ros::ok())
    {
        count++;
        cap >> img;
        if(!img.empty())
        {
            msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", img).toImageMsg();
            msg->header.stamp = ros::Time::now();
            pub.publish(msg);
            if(publish_camera_info)
            {
               caminfo_pub.publish(caminfo_msg); 
            }

            if(show_flag) 
            {
                namedWindow("camera", 0);
                imshow("camera", img);
                waitKey(1);
            }

            if(save_flag)
            {
                sprintf(imgname, "%05d.jpg", count);
                string str(imgname);
                imwrite(save_folder + str, img);
            }
        }

        // ROS_INFO( "Read the %dth frame successfully!", count );
    }

    // Exit
    cap.release();

    return 0;
}

File: src/video_capture.py
#! /usr/bin/env python

import os
import shutil
import rospy
from sensor_msgs.msg import Image
from sensor_msgs.msg import CameraInfo
from cv_bridge import CvBridge
import cv2


###--------------------------- Global variable ---------------------------###
bridge = CvBridge()

# parameter
save_folder = None
show_flag = None
save_flag = None
serial_port = None
publish_camera_info = None


###--------------------------- Internal funciton ---------------------------###
def GetCameraInfo():
    frame_id = rospy.get_param('~frame_id', 'camera')
    height = rospy.get_param('~height', 0)
    width = rospy.get_param('~width', 0)
    distortion_model = rospy.get_param('~distortion_model', '')
    D = rospy.get_param('~D')
    K = rospy.get_param('~K')
    R = rospy.get_param('~R')
    P = rospy.get_param('~P')
    binning_x = rospy.get_param('~binning_x', 0)
    binning_y = rospy.get_param('~binning_y', 0)
    roi_x_offset = rospy.get_param('~roi/x_offset', 0)
    roi_y_offset = rospy.get_param('~roi/y_offset', 0)
    roi_height = rospy.get_param('~roi/height', 0)
    roi_width = rospy.get_param('~roi/width', 0)
    roi_do_rectify = rospy.get_param<bool>('~roi/do_rectify', False)

    caminfo = CameraInfo()
    caminfo.header.frame_id = frame_id
    caminfo.height = height
    caminfo.width = width
    caminfo.distortion_model = distortion_model
    caminfo.D = D
    caminfo.K = K
    caminfo.R = R
    caminfo.P = P
    caminfo.binning_x = binning_x
    caminfo.binning_y = binning_y
    caminfo.roi.x_offset = roi_x_offset
    caminfo.roi.y_offset = roi_y_offset
    caminfo.roi.height = roi_height
    caminfo.roi.width = roi_width
    caminfo.roi.do_rectify = roi_do_rectify

    return caminfo


def video_capture():
    # Initial node
    rospy.init_node('video_capture', anonymous=True)
    rospy.loginfo('Start the video_capture node')

    # Parameter
    global save_folder, show_flag, save_flag, serial_port, publish_camera_info
    save_folder = rospy.get_param('~save_folder', None)
    show_flag = rospy.get_param('~show_flag', False)
    save_flag = rospy.get_param('~save_flag', False)
    serial_port = rospy.get_param('~serial_port', 0)
    publish_camera_info = rospy.get_param('~publish_camera_info', False)

    # Path
    if save_flag:
        if save_folder is None:
            rospy.logerr('save_folder is None!')
            return
        if os.path.exists(save_folder):
            shutil.rmtree(save_folder)
        os.makedirs(save_folder)

    # Camera parameter
    if publish_camera_info:
        caminfo_msg = GetCameraInfo()
        caminfo_pub = rospy.Publisher('camera_info', CameraInfo, queue_size=1)

    # Prepare to publish image topic
    img_pub = rospy.Publisher('image', Image, queue_size=1)

    # Open videocapture
    rospy.loginfo('serial_port: %s', serial_port)
    cap = cv2.VideoCapture(serial_port)
    if not cap.isOpened():
        rospy.logerr('Read video frame failed!')
        return

    # Read image
    count = 0
    while not rospy.is_shutdown():
        count += 1
        flag, img = cap.read()
        if flag:
            img_msg = bridge.cv2_to_imgmsg(img, 'bgr8')
            time = rospy.Time().now()
            img_msg.header.stamp = time
            img_pub.publish(img_msg)

            if publish_camera_info:
                caminfo_msg.header.stamp = time
                caminfo_pub.publish(caminfo_msg)

            if show_flag:
                cv2.namedWindow('image', 0)
                cv2.imshow('image', img)
                cv2.waitKey(1)

            if save_flag:
                filename = os.path.join(save_folder, '%06d.png' % count)
                cv2.imwrite(filename, img)

    # Exit
    cap.release()


if __name__ == '__main__':

    video_capture()
